Docker daemon crashes when pulling from a repository

Problem: 
When running a ‘docker pull’ against a Artifactory Instance hosting docker, the docker pull fails and the docker Daemon crashes.

Details:
This is running against the latest version of docker-machine with docker engine 1.8:

mikes-mbp:scripts mwalker$ docker --version
Docker version 1.8.1, build d12ea79
mikes-mbp:scripts mwalker$ docker-machine --version
docker-machine version 0.4.0 (9d0dc7a)
mikes-mbp:~ mwalker$ docker -D info
DEBU[0000] Trusting certs with subjects: [010U
mwalker]
Containers: 0
Images: 9
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 9
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.0.9-boot2docker
Operating System: Boot2Docker 1.8.1 (TCL 6.3); master : 7f12e95 - Thu Aug 13 03:24:56 UTC 2015
CPUs: 1
Total Memory: 2.959 GiB
Name: dev
ID: KVE2:QFFY:JU56:TEHP:R3XG:EXYL:ZOO7:DCWK:4JUS:C56Q:XSY4:UVB7
Debug mode (server): true
File Descriptors: 9
Goroutines: 19
System Time: 2015-08-19T21:34:37.519077883Z
EventsListeners: 0
Init SHA1:
Init Path: /usr/local/bin/docker
Docker Root Dir: /mnt/sda1/var/lib/docker
Username: mwalker638
Registry: https://index.docker.io/v1/
Labels:
 provider=virtualbox
mikes-mbp:~ mwalker$

Here is the docker pull command that fails:
mikes-mbp:scripts mwalker$  docker pull cos-tools-spoke:5002/tools-ca-com/git_spoke
Using default tag: latest
latest: Pulling from tools-ca-com/git_spoke
511136ea3c5a: Pulling fs layer
5b12ef8fd570: Pulling fs layer
8efe422e6104: Pulling fs layer
5b09b720e64d: Pulling fs layer
5a20c1f60aa2: Pulling fs layer
6b33827a0feb: Pulling fs layer
ec76391eca21: Pulling fs layer
754004d3c4d1: Pulling fs layer
ba5dff48af70: Pulling fs layer
2f7417f499ab: Pulling fs layer
ed9f2c19ddc2: Pulling fs layer
88e633c98112: Pulling fs layer
7b3b6fdee8a9: Pulling fs layer
9680b9670128: Pulling fs layer
9294744d544e: Pulling fs layer
4bb4a9f7d1e2: Pulling fs layer
74babb36f30d: Pulling fs layer
f32bf8a918e4: Pulling fs layer
5712e6e54cd2: Pulling fs layer
cda620ba24fb: Pulling fs layer
3d04e0458cbd: Pulling fs layer
390d6534d266: Pulling fs layer
724f1d4db774: Pulling fs layer
40715e4626dd: Pulling fs layer
54738c061619: Pulling fs layer
26e6227d4c10: Pulling fs layer
70a0ff31a573: Pulling fs layer
7175c700e781: Pulling fs layer
de066344c001: Pulling fs layer
5b4b9fffbd90: Pulling fs layer
23138238a27b: Pulling fs layer
38e0121ee391: Pulling fs layer
57b2d8b8d463: Pulling fs layer
599b9a2a57a9: Pulling fs layer
5345869f35a5: Pulling fs layer
e3d8a246303a: Pulling fs layer
e102bdecc99b: Pulling fs layer
c69ac56f14dc: Pulling fs layer
f76f218230ea: Pulling fs layer
2494799b2b72: Pulling fs layer
d97f7f7383b4: Pulling fs layer
39390046baae: Pulling fs layer
cc1ccca2ef54: Pulling fs layer
f8032c007c64: Pulling fs layer
6510e003cce9: Pulling fs layer
16bb05e43525: Pulling fs layer
f826f8d780c5: Pulling fs layer
d8a122e3bfa1: Pulling fs layer
f3ba9f78c0f7: Pulling fs layer
f025a1093332: Pulling fs layer
bee09c8a0a6f: Pulling fs layer
7b1f10892361: Pulling fs layer
9bd27e1fdff1: Pulling fs layer
7589059c8240: Pulling fs layer
dc5280db77f2: Pulling fs layer
fb344d3c650b: Pulling fs layer
b75580fd70b8: Pulling fs layer
aa41e0cbe291: Pulling fs layer
774b503c6d5b: Pulling fs layer
e8044d99bf93: Pulling fs layer
90a719a9b04f: Pulling fs layer
7233b14dab69: Pulling fs layer
c2e4ae3d5554: Pulling fs layer
0271d0bd6ef4: Pulling fs layer
7d2988ffb283: Pulling fs layer
2ae5a4fd0af6: Pulling fs layer
a1cc0f6567d0: Pulling fs layer
a8d3a1fb7e12: Layer already being pulled by another client. Waiting.
unexpected EOF
mikes-mbp:scripts mwalker$

Note – that this command also caused the docker daemon (running in boot2docker VM via docker-machine) to crash.  I find entries like this in the docker.log file:

time="2015-08-19T21:35:35.768528214Z" level=debug msg="pulling blob \"sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4\" t
time="2015-08-19T21:35:35.768607538Z" level=debug msg="pulling blob \"sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4\" t
time="2015-08-19T21:35:35.768684333Z" level=debug msg="pulling blob \"sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4\" t
time="2015-08-19T21:35:35.768761968Z" level=debug msg="pulling blob \"sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4\" t
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x38 pc=0xa77781]

goroutine 310 [running]:
github.com/docker/distribution/context.getLogrusLogger(0x0, 0x0, 0x0, 0x0, 0x0, 0xc208411720)
        /go/src/github.com/docker/docker/vendor/src/github.com/docker/distribution/context/logger.go:79 +0x91
github.com/docker/distribution/context.GetLogger(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/docker/docker/vendor/src/github.com/docker/distribution/context/logger.go:68 +0x62
github.com/docker/distribution/registry/storage/cache.(*cachedBlobStatter).Stat(0xc208173da0, 0x0, 0x0, 0xc208411720, 0x47, 0x0, 0x0, 0x0, 0x0, 0
        /go/src/github.com/docker/docker/vendor/src/github.com/docker/distribution/registry/storage/cache/cachedblobdescriptorstore.go:76 +0x316
github.com/docker/distribution/registry/client.(*blobs).Stat(0xc208a86000, 0x0, 0x0, 0xc208411720, 0x47, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/docker/docker/vendor/src/github.com/docker/distribution/registry/client/repository.go:384 +0xab
github.com/docker/docker/graph.(*v2Puller).download(0xc2082fc8c0, 0xc208047700)
        /go/src/github.com/docker/docker/graph/pull_v2.go:145 +0x7ef
created by github.com/docker/docker/graph.(*v2Puller).pullV2Tag
        /go/src/github.com/docker/docker/graph/pull_v2.go:269 +0x20c1

goroutine 1 [chan receive, 193 minutes]:
main.(*DaemonCli).CmdDaemon(0xc20802b0c0, 0xc20800a140, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/docker/docker/docker/daemon.go:292 +0x1811
…

The full docker.log file can be found here:

* https://dl.dropboxusercontent.com/u/17802995/artifactory-docker/logs-08-19-2015/docker.log

--- a/graph/pull_v2.go
+++ b/graph/pull_v2.go
@@ -141,9 +141,9 @@ func (p *v2Puller) download(di *downloadInfo) {
 	}
 	di.tmpFile = tmpFile
 
-	blobs := p.repo.Blobs(nil)
+	blobs := p.repo.Blobs(context.Background())
 
-	desc, err := blobs.Stat(nil, di.digest)
+	desc, err := blobs.Stat(context.Background(), di.digest)
 	if err != nil {
 		logrus.Debugf("Error statting layer: %v", err)
 		di.err <- err
@@ -151,7 +151,7 @@ func (p *v2Puller) download(di *downloadInfo) {
 	}
 	di.size = desc.Size
 
-	layerDownload, err := blobs.Open(nil, di.digest)
+	layerDownload, err := blobs.Open(context.Background(), di.digest)
 	if err != nil {
 		logrus.Debugf("Error fetching layer: %v", err)
 		di.err <- err

--- a/graph/push_v2.go
+++ b/graph/push_v2.go
@@ -138,7 +138,7 @@ func (p *v2Pusher) pushV2Tag(tag string) error {
 		dgst, err := p.graph.GetDigest(layer.ID)
 		switch err {
 		case nil:
-			_, err := p.repo.Blobs(nil).Stat(nil, dgst)
+			_, err := p.repo.Blobs(context.Background()).Stat(context.Background(), dgst)
 			switch err {
 			case nil:
 				exists = true
@@ -158,7 +158,7 @@ func (p *v2Pusher) pushV2Tag(tag string) error {
 		// if digest was empty or not saved, or if blob does not exist on the remote repository,
 		// then fetch it.
 		if !exists {
-			if pushDigest, err := p.pushV2Image(p.repo.Blobs(nil), layer); err != nil {
+			if pushDigest, err := p.pushV2Image(p.repo.Blobs(context.Background()), layer); err != nil {
 				return err
 			} else if pushDigest != dgst {
 				// Cache new checksum
@@ -226,7 +226,7 @@ func (p *v2Pusher) pushV2Image(bs distribution.BlobService, img *image.Image) (d
 
 	// Send the layer
 	logrus.Debugf("rendered layer for %s of [%d] size", img.ID, size)
-	layerUpload, err := bs.Create(nil)
+	layerUpload, err := bs.Create(context.Background())
 	if err != nil {
 		return "", err
 	}
@@ -250,7 +250,7 @@ func (p *v2Pusher) pushV2Image(bs distribution.BlobService, img *image.Image) (d
 	}
 
 	desc := distribution.Descriptor{Digest: dgst}
-	if _, err := layerUpload.Commit(nil, desc); err != nil {
+	if _, err := layerUpload.Commit(context.Background(), desc); err != nil {
 		return "", err
 	}
 

